lib/CustomFields/components/CustomFieldsForm/CustomFieldsForm.js

```diff
@@ -0,0 +1,361 @@
+import React, { Fragment, useReducer, useState, useMemo } from 'react';
+import { FormattedMessage } from 'react-intl';
+import PropTypes from 'prop-types';
+import {
+  isEqual,
+  uniqueId,
+} from 'lodash';
+
+import {
+  AccordionSet,
+  ExpandAllButton,
+  Paneset,
+  Pane,
+  Button,
+  PaneMenu,
+  PaneHeaderIconButton,
+  PaneFooter,
+  TextField,
+  Col,
+  Row,
+} from '@folio/stripes-components';
+
+import NavigationModal from '../../../Notes/components/NavigationModal';
+
+import FieldAccordion from '../FieldAccordion';
+
+import {
+  AddButton,
+  DeleteModal,
+} from './components';
+
+import {
+  fieldTypes,
+  fieldTypesLabels,
+  defaultFieldConfigs,
+} from '../../constants';
+
+import { textboxShape } from '../../shapes';
+
+import styles from './CustomFieldsForm.css';
+
+
+const propTypes = {
+  entityType: PropTypes.string.isRequired,
+  fetchUsageStatistics: PropTypes.func.isRequired,
+  initialValues: PropTypes.arrayOf(PropTypes.shape({
+    id: PropTypes.string.isRequired,
+    values: PropTypes.oneOfType([
+      textboxShape,
+    ]).isRequired,
+  })).isRequired,
+  onSubmit: PropTypes.func.isRequired,
+  redirectToView: PropTypes.func.isRequired,
+  submitSucceed: PropTypes.bool.isRequired,
+  submitting: PropTypes.bool.isRequired,
+};
+
+const CustomFieldsForm = ({
+  submitSucceed,
+  submitting,
+  fetchUsageStatistics,
+  redirectToView,
+  initialValues,
+  entityType,
+  onSubmit,
+}) => {
+  const getInitialAccordionsState = () => {
+    return initialValues.reduce((state, cf) => ({
+      ...state,
+      [cf.id]: false,
+    }), {});
+  };
+
+  const [accordions, setAccordions] = useState(getInitialAccordionsState);
+
+  const getInitialFormState = () => {
+    return {
+      customFields: initialValues,
+      pristine: true,
+      fieldsToDelete: [],
+    };
+  };
+
+  const stateReducer = (state, { type, payload }) => {
+    switch (type) {
+      case 'ADD_NEW_FIELD': {
```

Comment:
It would be great to store these actions in the constants and then use them here and in places where they are triggered.